### PR TITLE
Moved encryption-sdk from lib to modules to resolve dependency issues

### PR DESCRIPTION
**Pull request by vikasvb90**
### Description
To resolve dependency issues happened due to lib:encryption-sdk dependencies being pulled in core which was affecting other plugins, this PR is purposed to move crypto codebase to modules.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/crypto/CryptoHandlerRegistry.java
Comment (gbbafna) [inline server/src/main/java/org/opensearch/crypto/CryptoHandlerRegistry.java:60]: Please add a ToDo to allow this later based on the cluster settings .

File: server/src/main/java/org/opensearch/crypto/CryptoHandlerRegistry.java
Comment (gbbafna) [inline server/src/main/java/org/opensearch/crypto/CryptoHandlerRegistry.java:38]: replace `CryptoPlugin` with `CryptoPlugin<?,?> to avoid `Raw use of parameterized class 'CryptoPlugin' `

File: libs/encryption-sdk/licenses/aws-encryption-sdk-java-2.4.0.jar.sha1
Comment (gbbafna) [inline libs/encryption-sdk/licenses/aws-encryption-sdk-java-2.4.0.jar.sha1:1]: can we remove the folder as well ?

File: modules/crypto/src/main/java/org/opensearch/encryption/CryptoModulePlugin.java
Comment (gbbafna) [inline modules/crypto/src/main/java/org/opensearch/encryption/CryptoModulePlugin.java:62]: nit : pls rename to `createCryptoHandler` .

**Review by gbbafna — COMMENTED** [review]
Looks great. Some nit picks.  Thanks for restructuring it in such a short time.

File: modules/crypto/build.gradle
Comment (reta) [inline modules/crypto/build.gradle:18]: @vikasvb90 please move it to plugins (or even better, sandbox plugins), we should not install it by default, this is experimental feature

**Review by reta — CHANGES_REQUESTED** [review]

File: modules/crypto/build.gradle
Comment (gbbafna) [reply modules/crypto/build.gradle:18]: Thanks @reta  . Users need to provide `CryptoMetadata` while registering a repository. Hence this feature wouldn't get used by default.  The default installation would only populate the `CryptoHandlerRegistry` with one option of doing enc-dec . If we install it as a module , it makes usage  easier by just supplying/installing the KeyProvider  and not bothering about installing/configuring `CryptoModulePlugin`.

File: libs/encryption-sdk/licenses/aws-encryption-sdk-java-2.4.0.jar.sha1
Comment (vikasvb90) [reply libs/encryption-sdk/licenses/aws-encryption-sdk-java-2.4.0.jar.sha1:1]: This is a deleted file only.

File: modules/crypto/build.gradle
Comment (reta) [reply modules/crypto/build.gradle:18]: Thanks @gbbafna 

>  Hence this feature wouldn't get used by default. 

Modules supposed to have the functionality that is critical for OpenSearch to function, since this feature is optional (from above), the convince to use is not an argument to bundle it in. We could reconsider the decision(s) later on if needed.

Comment (reta) [reply modules/crypto/build.gradle:18]: In the same vein, this is experimental API and feature, why it is not behind the experimental flag?

Comment (vikasvb90) [reply modules/crypto/build.gradle:18]: @reta By this line ` since this feature is optional (from above), the convince to use is not an argument to bundle it in` are you saying that if there is a module then it becomes mandatory to be used in OpenSearch and it can't be a user's choice? I was looking at `repository-url` module and not sure if that fits your theory. Correct me pls if I am wrong!

Comment (reta) [reply modules/crypto/build.gradle:18]: >  are you saying that if there is a module then it becomes mandatory to be used in OpenSearch and it can't be a user's choice? I 

@vikasvb90 modules are installed automatically with each OpenSearch distribution, whereas plugins are bundled but not installed.

Comment (vikasvb90) [reply modules/crypto/build.gradle:18]: I am referring specifically to your point that a module once added to OpenSearch can't be disable or used optionally by a user. By this logic, `repository-url` should necessarily get used by all users. Right?

Comment (vikasvb90) [reply modules/crypto/build.gradle:18]: Intent here was to not treat it as optional or non-optional feature but as two flavors of data e.g. a repository. And that's why I think the decision was taken by @andrross  and @Bukhtawar . I will let others comment on this. But since the decision was already made, I would still prefer to push it now and then refactor later as we are a bit late to make the structural change.

Comment (reta) [reply modules/crypto/build.gradle:18]: @andrross @Bukhtawar I was reading through the discussion https://github.com/opensearch-project/OpenSearch/pull/8465#discussion_r1294940922 and I have not found any strong arguments toward making a module to support the cryptography (the name `crypto` is heavenly overloaded), any objections to have it as plugin since at this moment, it is not required and is purely opt-in

Comment (gbbafna) [reply modules/crypto/build.gradle:18]: Also, we want security to be first class citizen in OpenSearch and even preferred core for our changes . Due to issues in security plugin , we had to move this to module .

Comment (Bukhtawar) [reply modules/crypto/build.gradle:18]: @reta This isn't implementing the feature itself. The actual feature would have to be implemented by a concrete plugin implementation and hence I don't see much value with this logic to be made experimental as long as you see an upfront risk with this dependency which might cause regression with the existing code base.

Moving this to plugin would require two plugins to be installed which isn't the optimal experience IMO. Trying to understand your concern better

Comment (reta) [reply modules/crypto/build.gradle:18]: Thank @Bukhtawar 

> This isn't implementing the feature itself. The actual feature would have to be implemented by a concrete plugin implementation 

Now I am somewhat confused what is the role of this module? So we do have `CryptoPlugin` which is (to me) the extension point that has to be implemented by concrete plugin. This module is just another CryptoPlugin? Also, my biggest concern is module is apparently not tested (you could easily tell that by checking that there is no security policy and Bouncycastle needs tons of permissions).

Comment (vikasvb90) [reply modules/crypto/build.gradle:18]: `my biggest concern is module is apparently not tested` This has been perf tested. No jsm permission is required by AWS encryption SDK for it to be used in encryption/decryption. The blank which is nothing but NoOpCryptoHandler was added later as the review from crypto team is pending over partial encryption changes.

Comment (reta) [reply modules/crypto/build.gradle:18]: > This has been perf tested

@vikasvb90 how this plugin / module have been tested if this is the pull request that introduces it first time?

Comment (vikasvb90) [reply modules/crypto/build.gradle:18]: @reta I replied to this `there is no security policy and Bouncycastle needs tons of permissions`. I have perf tested this in lib and it didn't require any BC jsm permission.  I am about to test the restructured changes as well with actual CryptoHandler which does actual encryption/decryption.

Comment (vikasvb90) [reply modules/crypto/build.gradle:18]: Just recreated a draft PR again from previous commits - https://github.com/opensearch-project/OpenSearch/pull/9823 for your reference. This has actual encryption/decryption impl.

Comment (gbbafna) [reply modules/crypto/build.gradle:18]: The vision here is this module itself would contain all types of encryption handlers - frame based , block based , CTR  .  Depending upon the settings, a cluster would be able to chose its own method of encryption  .

Comment (reta) [reply modules/crypto/build.gradle:18]: @vikasvb90 thank you, since more details on the matter are revealed, I am more on the side to not even have changes delivered in 2.10 but move it off to 2.11 when with actual CryptoHandler which does actual encryption/decryption. If we want to show case uses this new feature, it is valid point - it should go to `plugins/example-plugins` instead, definitely (to me) not in a state to be promoted to a core  module.

Comment (reta) [reply modules/crypto/build.gradle:18]: @gbbafna thank you

> The vision here is this module itself would contain all types of encryption handlers - frame based , block based , CTR . Depending upon the settings, a cluster would be able to chose its own method of encryption .

Could we introduce this module / plugin / lib when vision becomes a bit more materialized? What value the current vision has for users?

**Review by peternied — CHANGES_REQUESTED** [review]
I'm cross referencing with the initial pull request for encryption-sdk [1], please forgive any misses between the PRs. I've also created separate threads for the different topics that are being discussed to make it easier to track resolution.

- [1] #9289

I think there are several areas that still need to be resolved before moving forward with this change, that are each captured in their own thread.
- Bouncy Castle Dependency
- Lib, Module, vs Plugin
- End to End Testing
- Security Review

File: modules/crypto/build.gradle
Comment (Bukhtawar) [reply modules/crypto/build.gradle:19]: Analyse/painless plugins can be optional based on the use cases. Not fully convinced that only critical features should sit in module.

This particular use case is solving the problem of not having end user worry about installing a plugin for handling the encryption logic and then another plugin for providing the keys

There were threads on making security a first class citizen. My understanding was we want to move to a model where security is enabled by default which makes modules a better choice

Comment (Bukhtawar) [reply modules/crypto/build.gradle:19]: cc: @reta @peternied

Comment (Bukhtawar) [reply modules/crypto/build.gradle:19]: > Thanks @Bukhtawar
> 
> > This particular use case is solving the problem of not having end user worry about installing a plugin for handling the encryption logic and then another plugin for providing the keys
> 
> This is not by exclusion, when we get to this stage - we could reconsider the decision to serve users more conveniently, now we are making the premature, unproven and unneeded decision that would be hard to reverse.


Would appreciate if you could take a look at the PR before assuming the same

File: modules/crypto/build.gradle
Comment (gbbafna) [reply modules/crypto/build.gradle:18]: @reta , would it be fine to push this as a  module for 2.11 then ? Meanwhile, we are removing encryption-sdk from core.

File: modules/crypto/build.gradle
Comment (reta) [inline modules/crypto/build.gradle:27]: Unrelated, but why we are using version that was release in 2020? https://mvnrepository.com/artifact/com.amazonaws/aws-encryption-sdk-java/1.7.0

Comment (vikasvb90) [reply modules/crypto/build.gradle:27]: https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/migration-guide.html Check this

File: modules/crypto/src/test/java/org/opensearch/encryption/CryptoModulePluginTests.java
Comment (vikasvb90) [reply modules/crypto/src/test/java/org/opensearch/encryption/CryptoModulePluginTests.java:22]: Doesn't make sense to me for NoOpCryptoHandler. They can get added during actual encryption/decryption.

File: modules/crypto/build.gradle
Comment (reta) [reply modules/crypto/build.gradle:27]: Please check this:

image

There are CVEs reported

Comment (vikasvb90) [reply modules/crypto/build.gradle:27]: First 2 are coming from bouncy castle and third one is coming from junit. I haven't used that dependency. I am using bcprov-jdk15to18 v1.75 which doesn't have a vulnerability.

Comment (vikasvb90) [reply modules/crypto/build.gradle:27]: First 2 are coming from bouncy castle (bcprov-ext-jdk15on) and third one is coming from junit. I haven't used that dependency. I am using bcprov-jdk15to18 v1.75 which doesn't have a vulnerability.
The direct vulnerability is actually a feature addition to strengthen the security. In order to adopt this, SDK needs to be migrated to 2.x. This can only be done if data encrypted by previous versions <1.7 have been ported to 1.7 first.

Comment (reta) [reply modules/crypto/build.gradle:27]: We don't have data encrypted with previous version <1.7, do we?

Comment (vikasvb90) [reply modules/crypto/build.gradle:27]: In OS we don't. If someone has customization done in BlobContainer implementations to encrypt data then they might be encrypting from previous versions.

Comment (reta) [reply modules/crypto/build.gradle:27]: > If someone has customization done in BlobContainer implementations to encrypt data then they might be encrypting from previous versions.

This is not the OpenSearch concern,  if someone has customization to encrypt, she definitely has the customization to decrypt. We should use the latest 2.4.1 I think.

File: modules/crypto/build.gradle
Comment (reta) [reply modules/crypto/build.gradle:18]: > @reta , would it be fine to push this as a module for 2.11 then ? Meanwhile, we are removing encryption-sdk from core : #9833

Thanks @gbbafna , we'll definitely get back to this once the 2.10 rush is over

File: modules/crypto/build.gradle
Comment (vikasvb90) [reply modules/crypto/build.gradle:27]: This is not about customization in decrypt because of customization in encrypt. This is about handling encrypted data of users who didn't have a choice then because OS didn't have encryption support and had to build their own encryption layer. OpenSearch bringing encryption now shouldn't force this on such users and just leave them with choice of either reindexing their data or ignoring the encryption altogether which might mean incompatible dependency issues in which case this option will no longer be valid. 

 It should be OpenSearch's concern to gracefully allow these users to migrate to newer versions especially when it can do so. AWS ESDK has v1.7 rolled out for this specific purpose and is both forward and backward compatible. OS can decide to upgrade SDK to 2.x in a major version upgrade (say 4.0) which would the right step aligned with major version bump property of introducing a breaking change.

Comment (reta) [reply modules/crypto/build.gradle:27]: @vikasvb90 we are talking about one specific library, it is not feasible to know, more to support, what kind of mechanisms or algorithms users could have been used for encryption / decryption of their data prior to whatever we would be introducing in OpenSearch. To your point,. we surely could help them and I have no issues supporting 1.x release line **if it is maintained**, so far it does not look like. People flood us with CVEs same day they are released, if this library + version is going to be flagged by security scanning tools (which it seems like it will) - this is no go.

**Comment by peternied** [comment]
@vikasvb90 How would you like to move forward with this pull request?

**Comment by vikasvb90** [comment]
> @vikasvb90 How would you like to move forward with this pull request?

@peternied I am closing this PR for now as we are yet to get approval from AWS Crypto SDK team on the customisations we have made to support multipart transfer.

**Pull request closed without merging**
